cpp/src/arrow/csv/writer.cc

```diff
@@ -72,11 +72,29 @@ RecordBatchIterator RecordBatchSliceIterator(const RecordBatch& batch,
   return RecordBatchIterator(std::move(functor));
 }
 
-// Counts the number of characters that need escaping in s.
-int64_t CountEscapes(util::string_view s) {
+// Counts the number of quotes in s.
+int64_t CountQuotes(util::string_view s) {
   return static_cast<int64_t>(std::count(s.begin(), s.end(), '"'));
 }
 
+// Counts the number of CSV structural characters in s.
+int64_t CountStructuralChars(util::string_view s) {
```

Comment:
```
int64_t CountStructuralChars2(const std::string& s) {
  int64_t result = 0;
  for (const auto c : s) {
      result += c == '"';
      result += c =='\n';
      result += c == '\r';
      result += c == ',';
  }
  return result;
}
```
Allows for auto-vectorization.

Reply:
Thinking about this a bit more we don't actually have to count. I moved this into `CheckStringHasNoStructuralChars` and use std::any_of there to detect if there are any structural chars in the value string.